unixfs/unixfs.go

```diff
@@ -258,3 +259,23 @@ func BytesForMetadata(m *Metadata) ([]byte, error) {
 func EmptyDirNode() *dag.ProtoNode {
 	return dag.NodeWithData(FolderPBData())
 }
+
+// ValidatePB validates a unixfs protonode.
+func ValidatePB(n *dag.ProtoNode, pb *pb.Data) error {
+	if len(pb.Blocksizes) == 0 { // special case
+		return nil
+	}
+	if len(n.Links()) != len(pb.Blocksizes) {
+		return errors.New("unixfs ill-formed, number of links does not match blocksize count")
+	}
+	total := uint64(len(pb.GetData()))
+	for _, blocksize := range pb.Blocksizes {
+		total += blocksize
+	}
+	if total != pb.GetFilesize() {
```

Comment:
So, one case here is `pb.Filesize == nil`. Is `Filesize` required?

Reply:
I double check, and if not I will add a check for that.

Reply:
Note: the function pb.GetFilesize() returns a uint64, if pb.Filesize == nil it will return 0

Reply:
Sorry, my question was more: If Filesize is nil, that means it wasn't specified. Do we *require* Filesize to be specified? Should we calculate it from the blocksizes if left unspecified?

Reply:
@Stebalien given there **must** be a source of size, and given that `blocksizes` are optional, my hunch would be to just make [`Filesize`](https://github.com/ipfs/go-ipfs/blob/v0.4.15/unixfs/pb/unixfs.proto#L15) hard-required.

Reply:
Hm. I'd like @whyrusleeping and @diasdavid to :+1: this.

@whyrusleeping, @diasdavid:

Will all IPFS files include a Filesize?

Reply:
I think its probably fine to assume that any correctly formatted ipfs/unixfs file will contain a filesize. Seems pretty hard to do a lot of things without it.

Reply:
Okay I added a test to make sure it is not nil.